src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,73 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static final int INITIAL_CAPACITY = 10;
+    private static final double SHRINK_THRESHOLD = 0.25;
+    private T[] elements;
+    private int size;
 
+    @SuppressWarnings("unchecked")
+    public MateStack() {
+        elements = (T[]) new Object[INITIAL_CAPACITY];
+        size = 0;
+    }
+
+    public void push(T value) {
+        ensureCapacity();
+        addValue(value);
     }
 
     public T peek() {
-        return null;
+        ensureNotEmpty();
+        return elements[size - 1];
     }
 
     public T pop() {
-        return null;
+        ensureNotEmpty();
+        return removeValue();
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void resize() {
+        T[] newArray = (T[]) new Object[elements.length * 2];
```

Comment:
Code Quality Issue: It would be a good idea to add a constant for the resize factor (2 in this case). This makes the code more readable and maintainable.

Reply:
Add resize factor as constant. Avoid using magic numbers: https://mate-academy.github.io/jv-program-common-mistakes/java-core/set-queue-stack-comparator/stack#avoid-magic-numbers-in-the-code